File: lib/evva/templates/swift/people_properties.swift
struct PropertyData {
	let name: String
	let value: Any
	let destinations: [Destination]

	init(name: String, value: Any, destinations: [Destination]) {
		self.name = name
		self.value = value
		self.destinations = destinations
	}

	init(type: PropertyType, value: Any) {
		self.init(name: type.name, value: value, destinations: type.destinations)
	}
}

enum PropertyType: String {
	<%- properties.each do |p| -%>
	case <%= p[:case_name] %> = "<%= p[:property_name ] %>"
	<%- end -%>

	var name: String { return rawValue }

	var destinations: [Destination] {
		switch self {
		<%- properties.each_with_index do |p, index| -%>
		case .<%= p[:case_name] %>: return [<%= p[:destinations].map { |d| ".#{d}" }.join(", ") %>]
		<%- end -%>
		}
	}
}

enum Property {
	<%- properties.each do |p| -%>
	case <%= p[:case_name] %>(<%= p[:type] %>)
	<%- end -%>

	var data: PropertyData {
		switch self {
		<%- properties.each_with_index do |p, index| -%>
		case let .<%= p[:case_name] %>(value):
			return PropertyData(type: .<%= p[:case_name] %>,
								value: value<% if p[:is_special_property] %>.rawValue<% end %>)
			<%- unless index == properties.count - 1 -%>

			<%- end -%>
		<%- end -%>
		}
	}
}

File: lib/evva/templates/kotlin/people_properties.kt
sealed class <%= class_name %>(
	property: <%= enums_class_name %>,
	val innerValue: Any? = null,
	val destinations: Array<<%= destinations_class_name %>> = emptyArray()
) {
	val name = property.key

	<%- properties.each_with_index do |property, index| -%>
	data class <%= property[:class_name] %>(
		val value: <%= property[:type] %>
	) : <%= class_name %>(
		property = <%= enums_class_name %>.<%= property[:property_name] %>,
		innerValue = value<% if property[:is_special_property] %>.key<% end %>,
		<%- if property[:destinations].count > 0 -%>
		destinations = arrayOf(
			<%- property[:destinations].each_with_index do |d, index| -%>
			<%= destinations_class_name %>.<%= d %><% if index < property[:destinations].count - 1 %>,<% end %>
			<%- end -%>
		)
		<%- end -%>
	)
	<%- unless index == properties.count - 1 -%>

	<%- end -%>
	<%- end -%>
}
